compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/IrInlineCodegen.kt

```diff
@@ -134,15 +132,19 @@ class IrExpressionLambdaImpl(
         ?: throw AssertionError("callable reference ${reference.dump()} has no name in context")
 
     override val capturedVars: List<CapturedParamDesc>
+    override val capturedVarsStart: Int
     override val invokeMethod: Method
     override val invokeMethodParameters: List<KotlinType?>
     override val invokeMethodReturnType: KotlinType
 
     init {
         val asmMethod = codegen.methodSignatureMapper.mapAsmMethod(function)
         val capturedParameters = reference.getArgumentsWithIr()
-        val captureStart = if (isExtensionLambda) 1 else 0 // extension receiver comes before captures
+        // Assume captured parameters form a sublist of all parameters (`function.explicitParameters`).
```

Comment:
Could you also describe a bit cases when it neccessery now. And I can';t understand why DI breaks it

Reply:
When InterfaceLowering (or something else, I don't know) moves the lambda from the interface to DefaultImpls, it transforms the extension receiver into a value parameter, so the assumption that captured value parameters are at the start of the list breaks.

The assumption that they are a sublist is only needed so that `capturedVarsStart` + `capturedVars` are enough to find them; you could always just store a list of indices to handle arbitrary gaps if you're willing to write more complicated logic in local remapping.

Reply:
>When InterfaceLowering (or something else, I don't know) moves the lambda from the interface to DefaultImpls, it transforms the extension receiver into a value parameter

It's strange. @gogabr @dnpetrov  do you know anything about it?

Reply:
next lambda is transformed `withO { this + k }`

Reply:
@pyos Could this code touch extension receiver https://github.com/JetBrains/kotlin/commit/99eab5a0582b37754478297bf7f7cde4978d6706#diff-a4a82dfcc1a970e31ab518f978ec8471bf221e600e3d7d0d21a1d58e38388b10R73 ?

Reply:
No, the purpose of this code is to transform non-local returns after LocalDeclarationsLowering has moved the lambdas wherever.

Reply:
@pyos Fix looks ok, but I'm trying to investigate where IR is corrupted. Looks like it's best place to fix

Reply:
Yes, indeed there `InterfaceLowering` affects lambda:
https://github.com/JetBrains/kotlin/commit/d1c2862e2722bc4e81e7f044c2bedac8c2ba2941
and 
https://github.com/JetBrains/kotlin/blob/efceb89b98a3a05a4ff1e1cb713541756309f3d5/compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt#L276